test/rgb_matrix/engine_test.exs

```diff
@@ -1,111 +1,235 @@
+      @test_pid {:pid, test_runner_pid}
+
+      @impl true
+      def new(_leds, _config), do: nil
+
+      @impl true
+      def render(_leds, _config), do: {:never, %{}, nil}
+
+      @impl true
+      def interact(_state, _config, led) do
+        {:pid, send_to} = @test_pid
+        send(send_to, {:interact, led})
+        {:ignore, nil}
+      end
+    end
+
+    [interact_animation: Animation.new(module_name, leds)]
+  end
+
+  defp add_leds(_context),
+    do: [leds: [Layout.LED.new(:l1, 0, 0)]]
+
+  # A paint_fn is necessary for regstration with the engine.
+  # This must be run as part of the setup for each test to make sure the pid
+  # is that of the test process.
+  defp add_paint_fn(_context) do
+    pid = self()
+
+    paint_fn = fn frame ->
+      send(pid, {:frame, frame})
+      :ok
+    end
+
+    [paint_fn: paint_fn]
+  end
+
+  # During unregistration, there is a chance the Engine could already be
+  # performing a render, or may have multiple :render messages in its mailbox,
+  # so it may send one or more frames to the given paintable anyway. So, we
+  # to receive and discard all frames.
+  defp maybe_receive_some_frames(frame) do
+    receive do
+      {:frame, ^frame} ->
+        maybe_receive_some_frames(frame)
+    after
+      100 ->
+        true
+    end
+  end
+
+  defp set_animation(%{animation: animation}),
+    do: :ok = Engine.set_animation(animation)
+
+  defmacrop assert_receive_down do
+    quote do
+      assert_receive {:DOWN, _ref, :process, _object, _reason}
+    end
+  end
+
+  defmacrop refute_receive_down do
+    quote do
+      refute_receive {:DOWN, _ref, :process, _object, _reason}
+    end
+  end
+
+  # Once for the entire test suite
+  setup_all [
+    :add_leds,
+    :add_frame,
+    :add_animation
+  ]
+
+  setup :add_paint_fn
+
+  describe "set_animation/1" do
+    @tag capture_log: true
+    test "crashes the engine when called with a non-animation" do
+      Process.monitor(Engine)
+
+      assert Engine.set_animation("We are") == :ok
+
+      assert_receive_down()
+    end
+
+    test "can set an animation", %{
+      animation: animation
+    } do
+      Process.monitor(Engine)
+
+      assert Engine.set_animation(animation) == :ok
+
+      refute_receive_down()
+    end
+  end
+
+  setup :set_animation
+
+  describe "register_paintable/1" do
+    @tag capture_log: true
+    test "crashes the engine when called with a non-function" do
+      Process.monitor(Engine)
+      fake_paint_fn = "the knights"
+
+      assert {:ok, ^fake_paint_fn, _frame} = Engine.register_paintable(fake_paint_fn)
+
+      assert_receive_down()
+    end
+
+    test "registers a paintable function", %{
+      paint_fn: paint_fn
+    } do
+      Process.monitor(Engine)
+
+      assert {:ok, ^paint_fn, _frame} = Engine.register_paintable(paint_fn)
+
+      refute_receive_down()
+    end
+  end
+
+  describe "renders frames" do
+    test "by calling registered paintables with the frame", %{
```

Comment:
⚠️ (recommended)

Since this is only one test, we probably don't need the `describe` block.

Reply:
Yeah, you're probably right.